kubelet/datadog_checks/kubelet/cadvisor.py

```diff
@@ -0,0 +1,177 @@
+        enabled_rates = instance.get('enabled_rates', DEFAULT_ENABLED_RATES)
+        self.enabled_rates = ["{0}.{1}".format(NAMESPACE, x) for x in enabled_rates]
+        pod_level_metrics = instance.get('pod_level_metrics', DEFAULT_POD_LEVEL_METRICS)
+        self.pod_level_metrics = ["{0}.{1}".format(NAMESPACE, x) for x in pod_level_metrics]
+
+        self._update_metrics(instance, cadvisor_url, pod_list, container_filter)
+
+    def _retrieve_cadvisor_metrics(self, cadvisor_url, timeout=10):
+        return requests.get(cadvisor_url, timeout=timeout).json()
+
+    def _update_metrics(self, instance, cadvisor_url, pod_list, container_filter):
+        metrics = self._retrieve_cadvisor_metrics(cadvisor_url)
+
+        if not metrics:
+            self.warning('cAdvisor returned no metrics')
+            return
+
+        for subcontainer in metrics:
+            c_id = subcontainer.get('id')
+            if 'aliases' not in subcontainer:
+                # it means the subcontainer is about a higher-level entity than a container
+                continue
+            try:
+                self._update_container_metrics(instance, subcontainer, pod_list, container_filter)
+            except Exception as e:
+                self.log.error("Unable to collect metrics for container: {0} ({1})".format(c_id, e))
+
+    def _publish_raw_metrics(self, metric, dat, tags, is_pod, depth=0):
+        """
+        Recusively parses and submit metrics for a given entity, until
+        reaching self.max_depth.
+        Nested metric names are flattened: memory/usage -> memory.usage
+        """
+        if depth >= self.max_depth:
+            self.log.warning('Reached max depth on metric=%s' % metric)
+            return
+
+        if isinstance(dat, numbers.Number):
+            # Pod level metric filtering
+            is_pod_metric = False
+            if self.pod_level_metrics and any([fnmatch(metric, pat) for pat in self.pod_level_metrics]):
+                is_pod_metric = True
+            if is_pod_metric != is_pod:
+                return
+
+            # Metric submission
+            if self.enabled_rates and any([fnmatch(metric, pat) for pat in self.enabled_rates]):
+                self.rate(metric, float(dat), tags)
+            elif self.enabled_gauges and any([fnmatch(metric, pat) for pat in self.enabled_gauges]):
+                self.gauge(metric, float(dat), tags)
+
+        elif isinstance(dat, dict):
+            for k, v in dat.iteritems():
+                self._publish_raw_metrics(metric + '.%s' % k.lower(), v, tags, is_pod, depth + 1)
+
+        elif isinstance(dat, list):
+            self._publish_raw_metrics(metric, dat[-1], tags, is_pod, depth + 1)
+
+    def _update_container_metrics(self, instance, subcontainer, pod_list, container_filter):
+        is_pod = False
+        in_static_pod = False
+        cid = subcontainer.get('id')
+        pod_uid = subcontainer.get('labels', []).get('io.kubernetes.pod.uid')
```

Comment:
Can the default be a dict instead of a list ? 
Especially with `.get('io.kubernetes.pod.uid')` over it (list doesn't support `get` on it)

Reply:
good catch, ouch